File: src/robcogen/cpp/generator.py
import os, logging, itertools
from types import SimpleNamespace as DictDot

from kgprim.ct.repr.mxrepr import MatrixRepresentation
from kgprim.ct.repr.mxrepr import MatrixReprMetadata
import kgprim.ct.repr.mxrepr as mxrepr

import ctgen
import ctgen.common
import ctgen_backends.cpp_iitrbd.generator as ctgencpp

import robcogen.luabridge as lua
import robcogen.core
import robcogen.vpc
import robcogen.utils.files as fileutils
import robcogen.cpp.config
import robcogen.cpp.templates

logger = logging.getLogger(__name__)


_path_here = fileutils.base_path(__file__)


class Generator:
    def __init__(self, robot, transforms, configurator, fileWriter):
        self.robot = robot
        self.transforms = transforms
        self.configurator = configurator
        self.fileWriter = fileWriter

        # The order of loading modules matter (sometimes), because some modules
        # define global symbols which are used by the other ones
        lua.lua_runtime.execute('generators = {}') # prepare a global
        self._loadLuaModule('utils.lua')
        self._loadLuaModule('common.lua')
        #self._loadLuaModule('tpl_fwd_dyn.lua')
        luat = self._loadLuaModule('generator.lua')

        self.generators = luat.generators(self.robot, transforms, configurator)

    def _loadLuaModule(self, filename):
        luaCodeSrc = open( os.path.join( _path_here, filename), "r")
        ret = lua.lua_runtime.execute(luaCodeSrc.read())
        luaCodeSrc.close()
        return ret

    def _genFile(self, ok, text, filename):
        if not ok :
            logger.error("Template evaluation failed (for file '{dest}'): {err}".format(dest=filename, err=text) )
        else :
            self.fileWriter.genFile( filename, text )

    def coreFiles(self):
        basename = self.configurator.files.h_types
        ok, text = self.generators.headers.types()
        self._genFile(ok, text, self.configurator.headerFileName(basename) )

        basename = self.configurator.files.h_main
        ok, text = self.generators.headers.main()
        self._genFile(ok, text, self.configurator.headerFileName(basename))

        basename = self.configurator.files.h_traits
        ok, text = self.generators.headers.traits()
        self._genFile(ok, text, self.configurator.headerFileName(basename))

        basename = self.configurator.files.h_constants
        ok, text = self.generators.constants.header()
        self._genFile(ok, text, self.configurator.headerFileName(basename) )
        ok, text = self.generators.constants.source()
        self._genFile(ok, text, self.configurator.implFileName(basename) )

        ok, text = self.generators.playground.source()
        self._genFile(ok, text, self.configurator.files.playground + '.cpp')


    def coordinateTransforms(self, ctModelMeta):
        aux = self.generators.constants.readAccessExprForGeometricConstant
        ctgenConfigurator = robcogen.cpp.config.CTGenConfigurator(
             self.robot.kinematics, self.configurator, ctModelMeta.ctModel, aux)

        ctgenerator = ctgencpp.Generator( ctgenConfigurator )

        matricesMetadata = self.transforms.allMatricesMetadata(MatrixRepresentation.homogeneous)

        (okh,header), (oks,source) = ctgenerator.generate_code(ctModelMeta,
                          {MatrixRepresentation.homogeneous : matricesMetadata})
        basename = self.configurator.files.h_transforms
        self._genFile(okh, header, self.configurator.headerFileName(basename) )
        self._genFile(oks, source, self.configurator.implFileName(basename) )


    def inertiaProperties(self):
        basename = self.configurator.files.h_inertia
        ok, text = self.generators.inertia.header()
        self._genFile(ok, text, self.configurator.headerFileName(basename))
        ok, text = self.generators.inertia.source()
        self._genFile(ok, text, self.configurator.implFileName(basename))

    def dynamics(self):
        basename = self.configurator.files.h_fwd_dyn
        ok, text = self.generators.fd.header()
        self._genFile(ok, text, self.configurator.headerFileName(basename))
        ok, text = self.generators.fd.source()
        self._genFile(ok, text, self.configurator.implFileName(basename))

        basename = self.configurator.files.h_inv_dyn
        ok, text = self.generators.id.header()
        self._genFile(ok, text, self.configurator.headerFileName(basename))
        ok, text = self.generators.id.source()
        self._genFile(ok, text, self.configurator.implFileName(basename))

        basename = self.configurator.files.h_jsim
        ok, text = self.generators.jsim.header()
        self._genFile(ok, text, self.configurator.headerFileName(basename))
        ok, text = self.generators.jsim.source()
        self._genFile(ok, text, self.configurator.implFileName(basename))

        basename = self.configurator.files.h_aba_hinv
        ok, text = self.generators.aba_hinv.header()
        self._genFile(ok, text, self.configurator.headerFileName(basename))
        ok, text = self.generators.aba_hinv.source()
        self._genFile(ok, text, self.configurator.implFileName(basename))

    def cmakefile(self):
        ok, text = self.generators.cmake.main()
        self._genFile(ok, text, 'CMakeLists.txt')

    def tests(self):
        originalWriter = self.fileWriter
        self.fileWriter = self.fileWriter.subDir('tests')

        ok, text = self.generators.tests.test_id()
        self._genFile(ok, text, self.configurator.files.test_cmdline_id+'.cpp')

        ok, text = self.generators.tests.test_jsim()
        self._genFile(ok, text, self.configurator.files.test_cmdline_jsim+'.cpp')

        ok, text = self.generators.tests.test_fd()
        self._genFile(ok, text, self.configurator.files.test_cmdline_fd+'.cpp')

        ok, text = self.generators.tests.test_consistency()
        self._genFile(ok, text, self.configurator.files.test_consistency+'.cpp')

        self.fileWriter = originalWriter

    def tests_for_installed_files(self):
        originalWriter = self.fileWriter
        self.fileWriter = self.fileWriter.subDir('tests_installation')

        ok, text = self.generators.tests.test_for_installed_files__consistency()
        self._genFile(ok, text, self.configurator.files.test_consistency+'.cpp')

        ok, text = self.generators.cmake.test_for_installed_files()
        self._genFile(ok, text, 'CMakeLists.txt')

        self.fileWriter = originalWriter

File: src/robcogen/cpp/generator.lua

local function pyIterableToTable(pobject)
    local ret = {}
    for v in python.iter(pobject) do
        table.insert(ret, v)
    end
    return ret
end


-- Use a local alias for the expected global modules
local cpputils  = RCG.cpp.utils
local common    = RCG.cpp.common
local genutils  = RCG.utils.templates
local iterutils = RCG.utils.iters


local function allGenerators(robot, transforms, configurator)
    local config    = configurator.txtCfg
    local cfgdata   = configurator.data
    local ns        = cpputils.ns_utils( config.namespaces(robot) )
    local ns_iitrbd = cpputils.ns_utils( pyIterableToTable(cfgdata.iitrbd.namespace) )

    local header_file_name = configurator.headerFileName
    local impl_file_name   = configurator.implFileName

    local env = {
        robot = robot,
        vars = config.vars,
        mxops = config.mxops,
        ns = ns,
        ns_iit_rbd = ns_iitrbd,
        types   = config.types,
        classes = config.classes,
        opts    = config.opts,
        meta    = config.meta,
        common = common(robot, transforms, configurator),
        utils = genutils,
        templateAll  = config.opts.template_all,
        includeGuard = config.internal.includeGuard(robot),
        typesMacro   = config.internal.typesMacro,
        headers = {
            main      = header_file_name(configurator.files.h_main),
            types     = header_file_name(configurator.files.h_types),
            constants = header_file_name(configurator.files.h_constants),
            traits    = header_file_name(configurator.files.h_traits),
            transforms= header_file_name(configurator.files.h_transforms),
            inertia   = header_file_name(configurator.files.h_inertia),
            fwd_dyn   = header_file_name(configurator.files.h_fwd_dyn),
            inv_dyn   = header_file_name(configurator.files.h_inv_dyn),
            jsim      = header_file_name(configurator.files.h_jsim),
        },
        impl_files = {
            constants = impl_file_name(configurator.files.h_constants),
            transforms= impl_file_name(configurator.files.h_transforms),
            inertia   = impl_file_name(configurator.files.h_inertia),
            inv_dyn   = impl_file_name(configurator.files.h_inv_dyn),
            fwd_dyn   = impl_file_name(configurator.files.h_fwd_dyn),
            jsim      = impl_file_name(configurator.files.h_jsim),
            playground= configurator.files.playground .. '.cpp',
            test_cmdline_id = configurator.files.test_cmdline_id .. '.cpp',
            test_cmdline_jsim = configurator.files.test_cmdline_jsim .. '.cpp',
            test_cmdline_fd = configurator.files.test_cmdline_fd .. '.cpp',
            test_consistency = configurator.files.test_consistency .. '.cpp',
        },
        sorted_links           = iterutils.get_sorted_links_iter_factory(robot),
        sorted_links_reversed  = iterutils.get_sorted_links_iter_factory(robot, true),
        sorted_joints          = iterutils.get_sorted_joints_iter_factory(robot),
        sorted_joints_reversed = iterutils.get_sorted_joints_iter_factory(robot, true),
        pairs = genutils.poly_pairs,
        ipairs= genutils.poly_ipairs,
        RCG = RCG,
    }

    env.leafs = {}
    for _, link in env.sorted_links() do
        if robot.treeutils.isLeaf(link) then
            table.insert(env.leafs, link)
        end
    end

    local generators_constants = RCG.cpp.generators.constants(robot, configurator, env)
    local ret = {
        common    = env.common, -- to let the caller inject something, possibly
        headers   = RCG.cpp.generators.core_headers(robot, configurator, env),
        constants = generators_constants,
        inertia   = RCG.cpp.generators.inertia_properties(robot, configurator, env, generators_constants.readAccessExprForInertiaConstant),
        id        = RCG.cpp.generators.inverse_dynamics(robot, configurator, env),
        jsim      = RCG.cpp.generators.jsim(robot, configurator, env),
        fd        = RCG.cpp.generators.forward_dynamics(robot, configurator, env),
        cmake     = RCG.cpp.generators.cmake(robot, configurator, env),
        tests     = RCG.cpp.generators.tests(robot, configurator, env),
        playground= RCG.cpp.generators.playground(robot, configurator, env),
    }

    return ret
end



return {
    generators = allGenerators
}
